packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp

```diff
@@ -0,0 +1,843 @@
+
+#include "MueLu_MasterList.hpp"
+#include "MueLu_Monitor.hpp"
+
+namespace MueLu {
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<const ParameterList> GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::GetValidParameterList() const {
+    RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+    //#define SET_VALID_ENTRY(name) validParamList->setEntry(name, MasterList::getEntry(name))
+    //SET_VALID_ENTRY("generalgeometric: coarsening rate");
+    //#undef  SET_VALID_ENTRY
+    // Coarsen can come in two forms, either a single char that will be interpreted as an integer which is used as the coarsening
+    // rate in every spatial dimentions, or it can be a longer string that will then be interpreted as an array of integers.
+    // By default coarsen is set as "2", hence a coarsening rate of 2 in every spatial dimension is the default setting!
+    validParamList->set< std::string>            ("Coarsen",                   "2", "Coarsening rate in all spatial dimensions");
+    validParamList->set< RCP<const FactoryBase> >("A",               Teuchos::null, "Generating factory of the matrix A");
+    validParamList->set< RCP<const FactoryBase> >("Nullspace",       Teuchos::null, "Generating factory of the nullspace");
+    validParamList->set< RCP<const FactoryBase> >("Coordinates",     Teuchos::null, "Generating factory for coorindates");
+
+    return validParamList;
+  }
+
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  void GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::DeclareInput(Level& fineLevel, Level& coarseLevel) const {
+    Input(fineLevel, "A");
+    Input(fineLevel, "Nullspace");
+
+    // check whether fine level coordinate information is available.
+    // If yes, request the fine level coordinates, coarse level coordinates
+    // will be generated by GeometricCoordinatesTransferFactory called by RAP factory
+    if (fineLevel.GetLevelID() == 0) {
+      if (fineLevel.IsAvailable("Coordinates", NoFactory::get())) {
+        fineLevel.DeclareInput("Coordinates", NoFactory::get(), this);
+      } else {
+	TEUCHOS_TEST_FOR_EXCEPTION(fineLevel.IsAvailable("Coordinates", NoFactory::get()),
+				   Exceptions::RuntimeError, "No Coordinates provided by the user.");
+      }
+    } else {
+      // on coarser levels we check the default factory providing "Coordinates"
+      // or the factory declared to provide "Coordinates".
+      // Then we check if the coordinates are available and eventually declare them as input.
+      RCP<const FactoryBase> myCoordsFact = GetFactory("Coordinates");
+      if (myCoordsFact == Teuchos::null) { myCoordsFact = fineLevel.GetFactoryManager()->GetFactory("Coordinates"); }
+      if (fineLevel.IsAvailable("Coordinates", myCoordsFact.get())) {
+        fineLevel.DeclareInput("Coordinates", myCoordsFact.get(), this);
+      }
+    }
```

Comment:
We have to break through a circular dependency: The GeneralGeometricPFactory needs fine level coordinates and calculates coarse coordinates. A Factory cannot request the data that it produces.